File: kin-core/src/androidTest/java/kin/core/FakeKinIssuer.java
package kin.core;


import static junit.framework.Assert.assertTrue;
import static kin.core.IntegConsts.TEST_NETWORK_URL;
import static kin.core.IntegConsts.URL_CREATE_ACCOUNT;

import java.io.IOException;
import java.util.concurrent.TimeUnit;
import okhttp3.OkHttpClient;
import okhttp3.Response;
import org.stellar.sdk.Asset;
import org.stellar.sdk.KeyPair;
import org.stellar.sdk.PaymentOperation;
import org.stellar.sdk.Server;
import org.stellar.sdk.Transaction;
import org.stellar.sdk.Transaction.Builder;
import org.stellar.sdk.responses.AccountResponse;
import org.stellar.sdk.responses.SubmitTransactionResponse;

/**
 * Fake issuer for integration test, support creating and funding accounts on stellar test net
 */
class FakeKinIssuer {

    private static final int TIMEOUT_SEC = 20;

    private final Server server;
    private final KeyPair issuerKeyPair;
    private final Asset kinAsset;

    FakeKinIssuer() throws IOException {
        this.server = new Server(TEST_NETWORK_URL, TIMEOUT_SEC, TimeUnit.SECONDS);
        this.issuerKeyPair = KeyPair.random();
        this.kinAsset = Asset.createNonNativeAsset("KIN", issuerKeyPair);
        createAndFundWithLumens(issuerKeyPair.getAccountId());
    }

    String getAccountId() {
        return issuerKeyPair.getAccountId();
    }

    private void createAndFundWithLumens(String accountId) throws IOException {
        OkHttpClient client = new OkHttpClient();
        okhttp3.Request request = new okhttp3.Request.Builder()
            .url(URL_CREATE_ACCOUNT + accountId).build();
        Response response = client.newCall(request).execute();
        assertTrue(response != null && response.body() != null && response.code() == 200);
    }

    void createAccount(String destinationAccount) throws Exception {
        createAndFundWithLumens(destinationAccount);
    }

    void fundWithKin(String destinationAccount, String amount) throws Exception {
        KeyPair destinationKeyPair = KeyPair.fromAccountId(destinationAccount);
        AccountResponse issuerAccountResponse = server.accounts().account(issuerKeyPair);
        PaymentOperation paymentOperation = new PaymentOperation.Builder(destinationKeyPair, kinAsset, amount)
            .build();
        Transaction transaction = new Builder(issuerAccountResponse)
            .addOperation(paymentOperation)
            .build();
        transaction.sign(issuerKeyPair);
        SubmitTransactionResponse response = server.submitTransaction(transaction);
        if (!response.isSuccess()) {
            throw Utils.createTransactionException(response);
        }
        assertTrue(response.isSuccess());
    }

}

File: kin-sdk/src/androidTest/kotlin/kin/sdk/FakeKinIssuer.kt
package kin.sdk


import android.util.Log
import kin.base.Asset
import kin.base.KeyPair
import kin.base.PaymentOperation
import kin.base.Server
import kin.base.Transaction
import kin.sdk.IntegConsts.TEST_NETWORK_URL
import kin.sdk.IntegConsts.URL_CREATE_ACCOUNT
import okhttp3.OkHttpClient
import org.junit.Assert.assertTrue
import java.io.IOException
import java.util.concurrent.TimeUnit

/**
 * Fake issuer for integration test, support creating and funding accounts on stellar test net
 */
internal class FakeKinIssuer @Throws(IOException::class)
constructor() {

    private val server: Server = Server(TEST_NETWORK_URL, TIMEOUT_SEC, TimeUnit.SECONDS)
    private val issuerKeyPair: KeyPair = KeyPair.random()
    private val kinAsset: Asset = Asset.createNonNativeAsset("KIN", issuerKeyPair)

    val accountId: String = issuerKeyPair.accountId

    init {
        createAndFundWithLumens(issuerKeyPair.accountId)
    }

    @Throws(IOException::class)
    private fun createAndFundWithLumens(accountId: String) {
        val client = OkHttpClient()
        val request = okhttp3.Request.Builder()
                .url(URL_CREATE_ACCOUNT + accountId).build()
        client.newCall(request).execute()?.let {
            if (it.body() == null || it.code() != 200) {
                Log.d("test", "createAndFundWithLumens error, error code = ${it.code()}, message = ${it.message()}")
            }
        }
    }

    @Throws(Exception::class)
    fun createAccount(destinationAccount: String) {
        createAndFundWithLumens(destinationAccount)
    }

    @Throws(Exception::class)
    fun fundWithKin(destinationAccount: String, amount: String) {
        val destinationKeyPair = KeyPair.fromAccountId(destinationAccount)
        val issuerAccountResponse = server.accounts().account(issuerKeyPair)
        val paymentOperation = PaymentOperation.Builder(destinationKeyPair, kinAsset, amount)
                .build()
        val transaction = Transaction.Builder(issuerAccountResponse)
                .addOperation(paymentOperation)
                .build()
        transaction.sign(issuerKeyPair)
        val response = server.submitTransaction(transaction)
        if (!response.isSuccess) {
            throw Utils.createTransactionException(response)
        }
        assertTrue(response.isSuccess)
    }
}

private val TIMEOUT_SEC = 20
